### PR TITLE
feat: add function to check if position is legally winnable for one side

### PR DESCRIPTION
**Pull request by gahtan-syarif**
needed for https://github.com/Disservin/fastchess/issues/410

### CODE REVIEW

**Comment by gahtan-syarif** [comment]
should i add this in chess.hpp too? @Disservin

**Comment by Disservin** [comment]
Rename to `hasSufficientMatingMaterial(Color color)` please  and then make `isInsufficientMaterial` use that + tests for the additional position. How to set it up is explained here https://github.com/Disservin/chess-library?tab=readme-ov-file#development-setup

**Comment by gahtan-syarif** [comment]
> Rename to `hasSufficientMatingMaterial(Color color)` please and then make `isInsufficientMaterial` use that + tests for the additional position. How to set it up is explained here https://github.com/Disservin/chess-library?tab=readme-ov-file#development-setup

i did meson compile -C build but its throwing a lot of errors?

**Comment by Disservin** [comment]
I cant help if I don't see the errors

**Comment by gahtan-syarif** [comment]
> I cant help if I don't see the errors

it shows this:
[error.txt](https://github.com/user-attachments/files/16740120/error.txt)
thats only some of it, the full one is thousands of lines long and basically the same stuff

**Comment by Disservin** [comment]
> > I cant help if I don't see the errors
> 
> it shows this: [error.txt](https://github.com/user-attachments/files/16740120/error.txt) thats only some of it, the full one is thousands of lines long and basically the same stuff

it aint compiling cause you fucked up the code xd

**Comment by Disservin** [comment]
https://github.com/Disservin/chess-library/pull/112/files#diff-9854d3303f96a5d1c4a1bcabfaf4f3e8275c50755d8a02c1fabf2ce69dc3657bL681

**Comment by gahtan-syarif** [comment]
ahh didnt see that lol

**Comment by gahtan-syarif** [comment]
ok i sucessfuly compiled it, what do i do now?

File: src/board.hpp [hunk 1]
```diff
@@ -680,38 +680,35 @@ class Board {
...
+                if (Square::same_color(bishops.lsb(), bishops.msb())) return false;
             }
         }
 
+        return true;
+    }
+
+    /// @brief Checks if the current position is a draw by insufficient material.
+    /// @return
+    [[nodiscard]] bool isInsufficientMaterial() const {
```
Comment (Disservin) [inline src/board.hpp:709]: just return `!hasSufficientMatingMaterial(Color::WHITE) && !hasSufficientMatingMaterial(Color::BLACK)`

**Comment by Disservin** [comment]
just add some test cases here https://github.com/Disservin/chess-library/blob/master/tests/board.cpp#L141 and then increment the 63 to 64 here https://github.com/Disservin/chess-library/blob/master/src/include.hpp and run the meson compile again

if you just want to test the board do
`meson test -C build --test-args="--test-suite=\"Board\""`

File: src/board.hpp (hunk 1 above)
Comment (gahtan-syarif) [reply src/board.hpp:709]: done 👍

**Comment by gahtan-syarif** [comment]
im getting this error
```
 ../tests/board.cpp:144:25: error: 'class chess::Board' has no member named 'hasSufficientMaterial'; did you mean 'isInsufficientMaterial'?
  144 |             CHECK(board.hasSufficientMaterial(board.sideToMove()) == false);
```

**Comment by gahtan-syarif** [comment]
> just add some test cases here https://github.com/Disservin/chess-library/blob/master/tests/board.cpp#L141 and then increment the 63 to 64 here https://github.com/Disservin/chess-library/blob/master/src/include.hpp and run the meson compile again
> 
> if you just want to test the board do `meson test -C build --test-args="--test-suite=\"Board\""`

done 👍 compile successfull

**Comment by Disservin** [comment]
> done 👍 compile successfull

i dont think so, otherwise there would be the chess.cpp file in git?

**Comment by gahtan-syarif** [comment]
> > done 👍 compile successfull
> 
> i dont think so, otherwise there would be the chess.cpp file in git?

ahh yeah, didnt realize it autoupdated my chess.hpp file, updated

**Comment by gahtan-syarif** [comment]
this pr is wrong, it seems there can be cases of mate when the king's own pieces block off the squares

**Pull request closed without merging**